shelley/chain-and-ledger/executable-spec/src/Cardano/Ledger/Value.hs

```diff
@@ -0,0 +1,258 @@
+-- is not stored in the Map, then its quantity is assumed to be 0.
+--
+-- Operations on assets are usually implemented 'pointwise'. That is, we apply
+-- the operation to the quantities for each asset in turn. So when we add two
+-- 'Value's the resulting 'Value' has, for each asset, the sum of the quantities
+-- of /that particular/ asset in the argument 'Value'. The effect of this is
+-- that the assets in the 'Value' are "independent", and are operated on
+-- separately.
+--
+-- We can think of 'Value' as a vector space whose dimensions are assets. At the
+-- moment there is only a single asset type (Ada), so 'Value' contains
+-- one-dimensional vectors. When asset-creating transactions are implemented,
+-- this will change and the definition of 'Value' will change to a 'Map Asset
+-- Int', effectively a vector with infinitely many dimensions whose non-zero
+-- values are recorded in the map.
+--
+-- To create a value of 'Value', we need to specifiy an asset policy. This can
+-- be done using 'Ledger.Ada.adaValueOf'. To get the ada dimension of 'Value' we
+-- use 'Ledger.Ada.fromValue'. Plutus contract authors will be able to define
+-- modules similar to 'Ledger.Ada' for their own assets.
+-- ============================================================================
+
+-- | Asset ID
+newtype AssetID = AssetID {assetID :: ByteString}
+  deriving newtype
+    ( Show,
+      Eq,
+      ToCBOR,
+      FromCBOR,
+      Ord,
+      NoUnexpectedThunks,
+      NFData
+    )
+
+-- | Policy ID
+newtype PolicyID era = PolicyID {policyID :: ScriptHash era}
+  deriving (Show, Eq, ToCBOR, FromCBOR, Ord, NoUnexpectedThunks, NFData)
+
+-- =========================================================================
+-- The Value type, and a few of its instances
+
+-- | The Value representing MultiAssets
+data Value era = Value !Integer !(Map.Map (PolicyID era) (Map.Map AssetID Integer))
+  deriving (Show, Generic)
+
+instance Eq (Value era) where
+  (Value c1 m1) == (Value c2 m2) = c1 == c2 && pointWise (==) m1 m2
+
+instance Typeable era => Monoid (Value era) where
+  mempty = Value 0 Map.empty
+
+instance Typeable era => Semigroup (Value era) where
+  (Value c1 m1) <> (Value c2 m2) = Value (c1 <+> c2) (m1 <+> m2)
+
+--TODO, FIX ME with some deriving magic
+instance NFData (Value era)
+
+-- TODO, FIX ME with some deriving magic
+instance NoUnexpectedThunks (Value era)
+
+-- CBOR
+instance
+  (Crypto era, Era era) =>
+  ToCBOR (Value era)
+  where
+  toCBOR (Value c v) =
+    encodeListLen 2
+      <> toCBOR c
+      <> toCBOR v
+
+instance
+  (Crypto era, Era era) =>
+  FromCBOR (Value era)
+  where
+  fromCBOR = do
+    decodeRecordNamed (fromString "Value") (const 2) $ do
+      c <- fromCBOR
+      v <- fromCBOR
+      pure $ Value c v
+
+-- =========================================================================
+-- The important Val instance
+
+instance (Era era, Crypto era, Typeable era) => Val (Value era) where
```

Comment:
The Val / Value naming is unfortunate, we need to find better names to clarify the distinction (not for this PR, but in a wider forum)

Reply:
I agree. One could make an argument for Token or Asset as these don't have widely known alternate meanings.